rundeckapp/grails-spa/packages/ui-trellis/src/app/components/activity/tests/savedFilters.spec.ts

```diff
@@ -0,0 +1,162 @@
+    });
+  });
+  afterEach(() => {
+    jest.clearAllMocks();
+  });
+  describe("Component Mounting", () => {
+    it("mounts properly with all props", () => {
+      expect(wrapper.exists()).toBe(true);
+    });
+    it("registers event listeners on mount", () => {
+      expect(eventBus.on).toHaveBeenCalledWith(
+        "invoke-save-filter",
+        expect.any(Function),
+      );
+    });
+    it("unregisters event listeners on unmount", () => {
+      // Create a mock event bus
+      const mockEventBus = {
+        on: jest.fn(),
+        off: jest.fn(),
+      };
+
+      // Mount the component
+      const wrapper = mount(SavedFilters, {
+        props: {
+          hasQuery: true,
+          query: {},
+          eventBus: mockEventBus,
+        },
+        global: {
+          mocks: {
+            $t: (msg) => msg,
+          },
+          stubs: {
+            btn: true,
+          },
+        },
+      });
+
+      const spy = jest.spyOn(mockEventBus, "off");
+
+      wrapper.unmount();
+
+      expect(spy).toHaveBeenCalledWith("invoke-save-filter");
+    });
+  });
+  describe("User Interactions", () => {
+    it("renders save button when hasQuery is true and query is empty", () => {
+      expect(wrapper.find('[data-test-id="save-filter-button"]').exists()).toBe(
+        true,
+      );
+    });
+    it("does not render delete filter button when query has no filterName", () => {
+      expect(wrapper.find('[data-test-id="delete-filter-btn"]').exists()).toBe(
+        false,
+      );
+    });
+    it("emits 'select_filter' when a filter is selected", async () => {
+      await wrapper.vm.selectFilter({ filterName: "Test Filter" });
+      expect(wrapper.emitted("select_filter")[0]).toEqual([
+        { filterName: "Test Filter" },
+      ]);
+    });
+    it("calls MessageBox.prompt when saveFilterPrompt is triggered", async () => {
+      await wrapper.vm.saveFilterPrompt();
+      expect(MessageBox.prompt).toHaveBeenCalled();
+    });
+  });
+  describe("Filter Management", () => {
+    it("updates filters after successful deletion", async () => {
+      await wrapper.vm.loadFilters();
+      await wrapper.vm.$nextTick();
+      const initialCount = wrapper.vm.filters.length;
+      if (initialCount > 0) {
+        console.log("Before delete:", wrapper.vm.filters);
+        // Mock the deleteFilter method
+        wrapper.vm.deleteFilter = (filterName) => {
+          const index = wrapper.vm.filters.findIndex(
+            (filter) => filter.filterName === filterName,
+          );
+          if (index !== -1) {
+            wrapper.vm.filters.splice(index, 1);
+          }
+        };
+        await wrapper.vm.deleteFilter(wrapper.vm.filters[0].filterName);
+        await wrapper.vm.$nextTick();
+        console.log("After delete:", wrapper.vm.filters);
+
+        expect(wrapper.vm.filters.length).toBeLessThan(initialCount);
+      } else {
+        throw new Error("No filters available to delete");
+      }
+    });
```

Comment:
A few things here:

- It's understandable if using if/else helps during the development of tests, but please avoid keeping them once tests are finished. Unless components have random functionalities, unit tests should be deterministic, which means that they use the same data every time and will always output the same results, therefore, there's no point in using if/else on them.

- instead of triggering wrapper.vm.deleteFilter, trigger the user interaction that will call the method deleteFilter;

- Instead of mocking the delete filter, mock the $confirm to return a resolved promise and either adjust `filterStore.removeFilter` to return the correct array of filters, OR before the click to trigger deleteFilter, do a mockReturnValue for `filterStore.loadForProject` to return an object with the filter property;

- lastly, instead of checking if filters.length is less than initialCount, it would be better to count the number of links rendered, from this block:

```
<li v-for="filter in filters" :key="filter.filterName">
          <a role="button" @click="selectFilter(filter)">
            {{ filter.filterName }}
            <span v-if="query && filter.filterName === query.filterName"
              >√</span
            >
          </a>
        </li>
```

To count them, it's possible to add a "data-test" attribute to the <a> tag, like this example [here](https://test-utils.vuejs.org/guide/essentials/a-crash-course#Making-the-test-pass) and then do the same logic of an initial count before removal, and a comparison of the numbers afterward.